addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1122,6 +1125,62 @@ func (r *ClusterBootstrapReconciler) updateValuesFromSecret(cluster *clusterapiv
 	return newSecret, nil
 }
 
+// cloneEmbeddedLocalObjectRef attempts to clone the embedded local object references from provider's namespace to cluster's
+// namespace. An example of embedded local object reference is the secret reference under CPIConfig.
+func (r *ClusterBootstrapReconciler) cloneEmbeddedLocalObjectRef(cluster *clusterapiv1beta1.Cluster, provider *unstructured.Unstructured) error {
+	groupKindNamesMap := util.ExtractTypedLocalObjectRef(provider.UnstructuredContent(), constants.LocalObjectRefSuffix)
+	if len(groupKindNamesMap) == 0 {
+		return nil
+	}
+
+	providerGVK := provider.GroupVersionKind()
+	r.Log.Info(fmt.Sprintf("cloning the embedded local object references within provider: %s with name: %s from"+
+		" %s namespace to %s namespace", provider.GroupVersionKind().String(), provider.GetName(), provider.GetNamespace(), cluster.Namespace))
+	for groupKind, resourceNames := range groupKindNamesMap {
+		gvr, err := r.getGVR(groupKind)
+		if err != nil {
+			// error has been logged within getGVR()
+			return err
+		}
+		for _, resourceName := range resourceNames {
+			r.Log.Info(fmt.Sprintf("cloning the GVR %s with name %s from %s namespace to %s namespace",
+				gvr.String(), resourceName, provider.GetNamespace(), cluster.Namespace))
+			fetchedObj, err := r.dynamicClient.Resource(*gvr).Namespace(provider.GetNamespace()).Get(r.context, resourceName, metav1.GetOptions{})
+			if err != nil {
+				r.Log.Error(err, fmt.Sprintf("unable to get provider: %s with name: %s under namespace: %s",
+					providerGVK.String(), provider.GetName(), provider.GetNamespace()))
+				return err
+			}
+
+			copiedObj := fetchedObj.DeepCopy()
+			unstructured.RemoveNestedField(copiedObj.Object, "metadata", "resourceVersion")
```

Comment:
Why are we removing resourceVersion?

Reply:
because `resoruceVersion` should not be set on object to be created. It is only used for concurrent control on updates. K8S errors out if you create an object with `resourceVersion` set.